### PR TITLE
Introduce benchmark suite for JSON reader options

### PR DESCRIPTION
**Pull request by shrshi**
## Description
The goal of this piece of work is to analyze the performance of the reader for JSON lines. This PR establishes a baseline for the performance of single quote normalization, white space normalization, mixed type as string parsing and recovery mode options when the input JSON is valid, and does not have any single quotes. 
Modifying the data generation to produce inputs with single quotes/mixed types/invalid lines will be the focus of follow-on PRs.
Addresses #15041 

## Checklist
- [X] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/benchmarks/io/json/json_reader_option.cpp [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+                                             static_cast<int32_t>(data_type::STRING),
+                                             static_cast<int32_t>(data_type::LIST),
+                                             static_cast<int32_t>(data_type::STRUCT)});
+
+  auto const tbl = create_random_table(
+    cycle_dtypes(data_types, num_cols), table_size_bytes{data_size}, data_profile_builder());
+  auto const view = tbl->view();
+  cudf::io::json_writer_options const write_opts =
+    cudf::io::json_writer_options::builder(source_sink.make_sink_info(), view)
+      .lines(true)
```
Comment (GregoryKimball) [inline cpp/benchmarks/io/json/json_reader_option.cpp:61]: While we are testing the options, I would recommend making `lines` into an nvbench enum axis as well.

**Comment by GregoryKimball** [comment]
Would you please post some early benchmark results? ([similar to this example](https://github.com/rapidsai/cudf/pull/15149#issuecomment-1967932079))

File: cpp/benchmarks/io/json/json_reader_option.cpp [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    nvbench::enum_type_list<normalize_single_quotes::NO, normalize_single_quotes::YES>,
+    nvbench::enum_type_list<mixed_types_as_string::NO, mixed_types_as_string::YES>,
+    nvbench::enum_type_list<recovery_mode::RECOVER_WITH_NULL, recovery_mode::FAIL>))
+  .set_name("json_read_all_lines")
+  .set_type_axes_names(
+    {"row_selection", "normalize_single_quotes", "mixed_types_as_string", "recovery_mode"})
+  .set_min_samples(6)
+  .add_int64_axis("num_chunks", {1})
+  .add_int64_axis("num_cols", {64})
+  .add_int64_power_of_two_axis("table_size", nvbench::range(26, 29, 1));
```
Comment (vuule) [inline cpp/benchmarks/io/json/json_reader_option.cpp:124]: is this temporary (to make those lovely charts), or planned to be merged?

**Comment by shrshi** [comment]
Benchmarks were run on A100 80GB GPU, with all combinations of the above options being enabled/disabled.
no_mixed_type
The figure above shows a slight drop in performance while byte range reading is enabled. Note that all the runs do not enable mixed_type_as_string option. 
mixed_type
Figure showing significant performance degradation when `mixed_type_as_string` is enabled.

File: cpp/benchmarks/io/json/json_reader_option.cpp (hunk 2 above)
Comment (shrshi) [reply cpp/benchmarks/io/json/json_reader_option.cpp:124]: I think the `table_size` or `num_cols` axes need not be merged, but I think we can retain the `num_chunks` (with some nvbench range) to showcase the byte range performance.

File: cpp/benchmarks/io/json/json_reader_option.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+      .na_rep("null")
+      .rows_per_chunk(100'000);
+  cudf::io::write_json(write_opts);
+
+  cudf::io::json_reader_options read_options =
+    cudf::io::json_reader_options::builder(source_sink.make_source_info())
+      .lines(json_lines_bool)
+      .normalize_single_quotes(false)
+      .normalize_whitespace(false)
+      .mixed_types_as_string(false);
```
Comment (vuule) [inline cpp/benchmarks/io/json/json_reader_option.cpp:61]: I think we're better off not including the options that are equal to the defaults.
```suggestion
      .lines(json_lines_bool);
```

File: cpp/benchmarks/io/json/json_reader_option.cpp [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+  auto mem_stats_logger = cudf::memory_stats_logger();
+  state.set_cuda_stream(nvbench::make_cuda_stream_view(cudf::get_default_stream().value()));
+  state.exec(
+    nvbench::exec_tag::sync | nvbench::exec_tag::timer, [&](nvbench::launch& launch, auto& timer) {
+      try_drop_l3_cache();
+      cudf::size_type num_rows_read = 0;
+      uint64_t num_cols_read        = 0;
```
Comment (vuule) [inline cpp/benchmarks/io/json/json_reader_option.cpp:69]:
```suggestion
      cudf::size_type num_cols_read = 0;
```

File: cpp/benchmarks/io/json/json_reader_option.cpp [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+  size_t const chunk_size = cudf::util::div_rounding_up_safe(source_sink.size(), num_chunks);
+  auto mem_stats_logger   = cudf::memory_stats_logger();
+  state.set_cuda_stream(nvbench::make_cuda_stream_view(cudf::get_default_stream().value()));
+  state.exec(
+    nvbench::exec_tag::sync | nvbench::exec_tag::timer, [&](nvbench::launch& launch, auto& timer) {
+      try_drop_l3_cache();
+      cudf::size_type num_rows_read = 0;
+      uint64_t num_cols_read        = 0;
+      timer.start();
+      switch (RowSelection) {
```
Comment (vuule) [inline cpp/benchmarks/io/json/json_reader_option.cpp:149]: should we have num_rows/skiprows here as well?

File: cpp/benchmarks/io/json/json_reader_option.cpp [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+
+// Size of the data in the benchmark dataframe; chosen to be low enough to allow benchmarks to
+// run on most GPUs, but large enough to allow highest throughput
+constexpr size_t data_size         = 512 << 20;
+constexpr cudf::size_type num_cols = 64;
+
+template <json_lines JsonLines>
+void BM_json_read_options(nvbench::state& state, nvbench::type_list<nvbench::enum_type<JsonLines>>)
+{
+  auto constexpr json_lines_bool = JsonLines == json_lines::YES;
```
Comment (vuule) [inline cpp/benchmarks/io/json/json_reader_option.cpp:36]: Nit: I think we (more often) use `constexpr auto`. Admittedly, not very consistently across the code base.

File: cpp/benchmarks/io/json/json_reader_option.cpp [hunk 7]
```diff
@@ -0,0 +1,203 @@
...
+          break;
+        }
+        case row_selection::BYTE_RANGE: {
+          for (uint64_t chunk = 0; chunk < num_chunks; chunk++) {
+            read_options.set_byte_range_offset(chunk * chunk_size);
+            read_options.set_byte_range_size(chunk_size);
+            auto const result = cudf::io::read_json(read_options);
+            num_rows_read += result.tbl->num_rows();
+            num_cols_read = result.tbl->num_columns();
+            if (num_cols_read)
```
Comment (vuule) [inline cpp/benchmarks/io/json/json_reader_option.cpp:163]: when is this false? :cold_sweat:

**Review by vuule — COMMENTED** [review]
Great stuff!
Some minor comments/questions.

File: cpp/benchmarks/io/json/json_reader_option.cpp (hunk 7 above)
Comment (shrshi) [reply cpp/benchmarks/io/json/json_reader_option.cpp:163]: When no rows are read in a particular chunk. I spent some time wrestling with this bug on Friday until I realized (after our offline discussion) that if there are no rows starting in a chunk, the byte range reader returns 0 rows and 0 columns. So the `CUDF_EXPECTS` assertion is moved to inside the chunk iterations.

Comment (vuule) [reply cpp/benchmarks/io/json/json_reader_option.cpp:163]: I see. But, how come there's a chunk with no rows? looks like we read in up to five chunks.

Comment (shrshi) [reply cpp/benchmarks/io/json/json_reader_option.cpp:163]: When `num_chunks=5` the last chunk does not have any rows starting in it. The assertion failed there earlier.

Comment (vuule) [reply cpp/benchmarks/io/json/json_reader_option.cpp:163]: from an offline discussion: there's an issue with the data generator leading to huge last rows in lists columns. I will look into this.
In the meantime, this check can stay in this PR.

**Comment by shrshi** [comment]
After updates to the mixed types as string handling and the byte range reader, here are the updated benchmarks.
rsALL_rmFAIL
`row_selection=ALL, recovery_mode=FAIL`


rsALL_rmNULL
`row_selection=ALL, recovery_mode=RECOVER_WITH_NULL`


rsBYTE_rmFAIL
`row_selection=BYTE_RANGE, recovery_mode=FAIL`


rsBYTE_rmNULL
`row_selection=BYTE_RANGE, recovery_mode=FAIL`

**Comment by GregoryKimball** [comment]
These plots are great! Please check out https://github.com/rapidsai/cudf/issues/15185 for an example of where the byte range support is not working optimally.

File: cpp/benchmarks/io/json/json_reader_option.cpp (hunk 5 above)
Comment (shrshi) [reply cpp/benchmarks/io/json/json_reader_option.cpp:145]: I couldn't find a skip_rows option in the json reader options. I think we will have to do the row offsetting using byte ranges to test numrows, but we are already testing that.

Comment (vuule) [reply cpp/benchmarks/io/json/json_reader_option.cpp:145]: omg, read_json does not have num_rows/skiprows. great news :)

**Review by vuule — APPROVED** [review]

File: cpp/benchmarks/io/json/json_reader_option.cpp [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+  state.set_cuda_stream(nvbench::make_cuda_stream_view(cudf::get_default_stream().value()));
+  state.exec(
+    nvbench::exec_tag::sync | nvbench::exec_tag::timer, [&](nvbench::launch& launch, auto& timer) {
+      try_drop_l3_cache();
+      cudf::size_type num_rows_read = 0;
+      cudf::size_type num_cols_read = 0;
+      timer.start();
+      auto const result = cudf::io::read_json(read_options);
+      num_rows_read     = result.tbl->num_rows();
+      num_cols_read     = result.tbl->num_columns();
```
Comment (PointKernel) [inline cpp/benchmarks/io/json/json_reader_option.cpp:69]:
```suggestion
      timer.start();
      auto const result = cudf::io::read_json(read_options);
      auto const num_rows_read     = result.tbl->num_rows();
      auto const num_cols_read     = result.tbl->num_columns();
```

**Review by PointKernel — APPROVED** [review]
One non-blocking nit.

Very clean code, nice work!

File: cpp/benchmarks/io/json/json_reader_option.cpp [hunk 9]
```diff
@@ -0,0 +1,197 @@
...
+  state.add_buffer_size(
+    mem_stats_logger.peak_memory_usage(), "peak_memory_usage", "peak_memory_usage");
+  state.add_buffer_size(source_sink.size(), "encoded_file_size", "encoded_file_size");
+}
+
+template <row_selection RowSelection,
+          normalize_single_quotes NormalizeSingleQuotes,
+          normalize_whitespace NormalizeWhitespace,
+          mixed_types_as_string MixedTypesAsString,
+          recovery_mode RecoveryMode>
```
Comment (ttnghia) [inline cpp/benchmarks/io/json/json_reader_option.cpp:85]: Oh no this is too many template params. Why not using run time parameter instead?

Comment (shrshi) [reply cpp/benchmarks/io/json/json_reader_option.cpp:85]: I followed the design for benchmarking reader options in [orc](https://github.com/rapidsai/cudf/blob/6c31eacce54d8f1c9d3ad183d8879e09e679b369/cpp/benchmarks/io/orc/orc_reader_options.cpp#L47) and [parquet](https://github.com/rapidsai/cudf/blob/6c31eacce54d8f1c9d3ad183d8879e09e679b369/cpp/benchmarks/io/parquet/parquet_reader_options.cpp#L47). Would we have to modify those benchmarks as well to maintain similar design?

Comment (PointKernel) [reply cpp/benchmarks/io/json/json_reader_option.cpp:85]: >Oh no this is too many template params. Why not using run time parameter instead? 

This makes results more readable and the build time for benchmarks doesn't really matter IMO.

Comment (ttnghia) [reply cpp/benchmarks/io/json/json_reader_option.cpp:85]: Yeah that's fine to me :+1:

**Review by ttnghia — APPROVED** [review]

**Pull request merged**